{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4032fe7b-893e-42b1-a6c2-06495aec3302",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.core import preprocess_generate_masks\n",
    "%matplotlib inline\n",
    "\n",
    "# @title Preprocessing Settings { display-mode: \"form\" }\n",
    "src = \"path\"  # @param {type:\"string\"}  # (path) path to source folder (where original images were stored)\n",
    "metadata_type = \"cellvoyager\"  # @param [\"cellvoyager\", \"cq1\", \"Nikon\"] {type:\"string\"}  # (string) - type of filename metadata\n",
    "custom_regex = None  # @param {type:\"raw\"}  # (regex) - Regular expression if filename metadata not in metadata_type\n",
    "experiment = \"screen\"  # @param {type:\"string\"}  # (string) - Name of experiment\n",
    "channels = [0, 1, 2, 3]  # @param {type:\"raw\"}  # (list) - list of integers representing available channels\n",
    "\n",
    "cell_channel = 3  # @param {type:\"integer\"}  # (integer or NoneType) - Cell image dimension\n",
    "cell_background = 100  # @param {type:\"integer\"}  # (integer) - Background value in cell images\n",
    "cell_Signal_to_noise = 10  # @param {type:\"integer\"}  # (integer) - Signal to noise ratio for cell channel\n",
    "cell_CP_prob = -1  # @param {type:\"integer\"}  # (integer) - Cellpose Cell probability\n",
    "remove_background_cell = False  # @param {type:\"boolean\"}  # (bool) - Set background to 0 for cell channel\n",
    "\n",
    "nucleus_channel = 0  # @param {type:\"integer\"}  # (Optional, integer or NoneType) - Nucleus image dimension\n",
    "nucleus_background = 200  # @param {type:\"integer\"}  # (Optional, integer) - Background value in nucleus images\n",
    "nucleus_Signal_to_noise = 5  # @param {type:\"integer\"}  # (Optional, integer) - Signal to noise ratio for nucleus channel\n",
    "nucleus_CP_prob = 0  # @param {type:\"integer\"}  # (Optional, integer) - Cellpose Nucleus probability\n",
    "remove_background_nucleus = False  # @param {type:\"boolean\"}  # (Optional, bool) - Set background to 0 for nucleus channel\n",
    "\n",
    "pathogen_model = None  # @param {type:\"raw\"}  # (Optional, path or NoneType) - Custom cellpose model path for pathogen detection\n",
    "pathogen_channel = 2  # @param {type:\"integer\"}  # (Optional, integer or NoneType) - Pathogen image dimension\n",
    "pathogen_background = 150  # @param {type:\"integer\"}  # (Optional, integer) - Background value in pathogen images\n",
    "pathogen_Signal_to_noise = 6  # @param {type:\"integer\"}  # (Optional, integer) - Signal to noise ratio for pathogen channel\n",
    "pathogen_CP_prob = -2  # @param {type:\"integer\"}  # (Optional, integer) - Cellpose pathogen probability\n",
    "remove_background_pathogen = True  # @param {type:\"boolean\"}  # (Optional, bool) - Set background to 0 for pathogen channel\n",
    "\n",
    "consolidate = False  # @param {type:\"boolean\"}  # (bool) - Consolidate overlapping objects\n",
    "magnification = 20  # @param [20, 40, 60, 100] {type:\"raw\"}  # (integer) - Objective magnification used\n",
    "save = True  # @param {type:\"boolean\"}  # (bool) - Save masks and object data to database\n",
    "preprocess = True  # @param {type:\"boolean\"}  # (bool) - Preprocess images\n",
    "masks = True  # @param {type:\"boolean\"}  # (bool) - Generate masks\n",
    "batch_size = 100  # @param {type:\"integer\"}  # (integer) - Number of images to be normalized together\n",
    "filter = False  # @param {type:\"boolean\"}  # (bool) - Filter objects based on size\n",
    "merge_pathogens = False  # @param {type:\"boolean\"}  # (bool) - Merge pathogens that share > 75% perimeter\n",
    "plot = False  # @param {type:\"boolean\"}  # (bool) - Plot normalized intensity and object images\n",
    "adjust_cells = True  # @param {type:\"boolean\"}  # (bool) - If cell, nucleus and pathogen: merge cells that share a pathogen\n",
    "test_mode = False  # @param {type:\"boolean\"}  # (bool) - Test settings in test mode before analyzing entire experiment\n",
    "test_images = 100  # @param {type:\"integer\"}  # (integer) - Number of images to analyze in test mode\n",
    "random_test = True  # @param {type:\"boolean\"}  # (bool) - Randomize images for test mode\n",
    "\n",
    "settings = {\n",
    "    'src': src,\n",
    "    'metadata_type': metadata_type,\n",
    "    'custom_regex': custom_regex,\n",
    "    'experiment': experiment,\n",
    "    'channels': channels,\n",
    "    'cell_channel': cell_channel,\n",
    "    'cell_background': cell_background,\n",
    "    'cell_Signal_to_noise': cell_Signal_to_noise,\n",
    "    'cell_CP_prob': cell_CP_prob,\n",
    "    'remove_background_cell': remove_background_cell,\n",
    "    'nucleus_channel': nucleus_channel,\n",
    "    'nucleus_background': nucleus_background,\n",
    "    'nucleus_Signal_to_noise': nucleus_Signal_to_noise,\n",
    "    'nucleus_CP_prob': nucleus_CP_prob,\n",
    "    'remove_background_nucleus': remove_background_nucleus,\n",
    "    'pathogen_model': pathogen_model,\n",
    "    'pathogen_channel': pathogen_channel,\n",
    "    'pathogen_background': pathogen_background,\n",
    "    'pathogen_Signal_to_noise': pathogen_Signal_to_noise,\n",
    "    'pathogen_CP_prob': pathogen_CP_prob,\n",
    "    'remove_background_pathogen': remove_background_pathogen,\n",
    "    'consolidate': consolidate,\n",
    "    'magnification': magnification,\n",
    "    'save': save,\n",
    "    'preprocess': preprocess,\n",
    "    'masks': masks,\n",
    "    'batch_size': batch_size,\n",
    "    'filter': filter,\n",
    "    'merge_pathogens': merge_pathogens,\n",
    "    'plot': plot,\n",
    "    'adjust_cells': adjust_cells,\n",
    "    'test_mode': test_mode,\n",
    "    'test_images': test_images,\n",
    "    'random_test': random_test\n",
    "}\n",
    "\n",
    "preprocess_generate_masks(settings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74bf02d1-9f7c-4ac2-82d6-68b3a66d19dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.measure import measure_crop\n",
    "%matplotlib inline\n",
    "\n",
    "# @title Measurement & Cropping Settings { display-mode: \"form\" }\n",
    "\n",
    "# === General ===\n",
    "src = \"path\"  # @param {type:\"string\"}  # (path) Path to source folder (ends in /merged)\n",
    "channels = [0, 1, 2, 3]  # @param {type:\"raw\"}  # (list) - Available channel indices\n",
    "\n",
    "# === Object mask dimensions ===\n",
    "cell_mask_dim = 4  # @param {type:\"integer\"}  # (int or None) - Cell mask dimension\n",
    "nucleus_mask_dim = 5  # @param {type:\"integer\"}  # (int or None) - Nucleus mask dimension\n",
    "pathogen_mask_dim = 6  # @param {type:\"integer\"}  # (int or None) - Pathogen mask dimension\n",
    "\n",
    "# === Object size filters ===\n",
    "cell_min_size = 2000  # @param {type:\"integer\"}  # (int) - Minimum cell area (px²)\n",
    "nucleus_min_size = 1000  # @param {type:\"integer\"}  # (int) - Minimum nucleus area (px²)\n",
    "pathogen_min_size = 400  # @param {type:\"integer\"}  # (int) - Minimum pathogen area (px²)\n",
    "cytoplasm_min_size = 0  # @param {type:\"integer\"}  # (int) - Minimum cytoplasm area (px²)\n",
    "\n",
    "# === Cropping settings ===\n",
    "crop_mode = [\"cell\"]  # @param {type:\"raw\"}  # (list) - Objects to crop into PNGs ('cell', 'nuclei', 'pathogen')\n",
    "use_bounding_box = False  # @param {type:\"boolean\"}  # (bool) - Use bounding box instead of exact object mask\n",
    "png_size = [[224, 224]]  # @param {type:\"raw\"}  # (list of lists) - PNG image size\n",
    "png_dims = [0, 1, 2]  # @param {type:\"raw\"}  # (list) - Channels to include in PNGs\n",
    "normalize = False  # @param {type:\"raw\"}  # (bool or list) - Normalize PNGs to percentiles\n",
    "normalize_by = \"png\"  # @param [\"png\", \"fov\"] {type:\"string\"}  # (str) - Normalize by 'png' or 'fov'\n",
    "\n",
    "# === Saving and plotting ===\n",
    "save_png = True  # @param {type:\"boolean\"}  # (bool) - Save PNGs\n",
    "save_measurements = True  # @param {type:\"boolean\"}  # (bool) - Save object measurements\n",
    "plot = False  # @param {type:\"boolean\"}  # (bool) - Plot images during analysis\n",
    "plot_filtration = False  # @param {type:\"boolean\"}  # (bool) - Plot filtration steps\n",
    "uninfected = False  # @param {type:\"boolean\"}  # (bool) - Include uninfected objects\n",
    "\n",
    "# === Test mode ===\n",
    "test_mode = False  # @param {type:\"boolean\"}  # (bool) - Activate test mode\n",
    "test_nr = 10  # @param {type:\"integer\"}  # (int) - Number of images to analyze in test mode\n",
    "\n",
    "# === Assemble settings dictionary and run ===\n",
    "settings = {\n",
    "    'src': src,\n",
    "    'channels': channels,\n",
    "    'cell_mask_dim': cell_mask_dim,\n",
    "    'cell_min_size': cell_min_size,\n",
    "    'nucleus_mask_dim': nucleus_mask_dim,\n",
    "    'nucleus_min_size': nucleus_min_size,\n",
    "    'pathogen_mask_dim': pathogen_mask_dim,\n",
    "    'pathogen_min_size': pathogen_min_size,\n",
    "    'cytoplasm_min_size': cytoplasm_min_size,\n",
    "    'save_png': save_png,\n",
    "    'crop_mode': crop_mode,\n",
    "    'use_bounding_box': use_bounding_box,\n",
    "    'png_size': png_size,\n",
    "    'normalize': normalize,\n",
    "    'png_dims': png_dims,\n",
    "    'normalize_by': normalize_by,\n",
    "    'save_measurements': save_measurements,\n",
    "    'plot': plot,\n",
    "    'plot_filtration': plot_filtration,\n",
    "    'uninfected': uninfected,\n",
    "    'test_mode': test_mode,\n",
    "    'test_nr': test_nr\n",
    "}\n",
    "\n",
    "measure_crop(settings)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38090c27-49d5-44ee-835a-ae5e04dc9194",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.ml import generate_ml_scores\n",
    "%matplotlib inline\n",
    "\n",
    "# @title ML Training Settings { display-mode: \"form\" }\n",
    "\n",
    "# === Data and model configuration ===\n",
    "src = \"path\"  # @param {type:\"string\"}  # (path) Path to source folder (where original images were stored)\n",
    "model_type_ml = \"xgboost\"  # @param [\"random_forest\", \"xgboost\", \"gradient_boosting\"] {type:\"string\"}  # (str) Type of model\n",
    "n_estimators = 100  # @param {type:\"integer\"}  # (int) Number of estimators in the model\n",
    "test_size = 0.2  # @param {type:\"number\"}  # (float) Fraction of data used for the test set\n",
    "\n",
    "# === Controls and grouping ===\n",
    "location_column = \"column_name\"  # @param {type:\"string\"}  # (str) Column with control metadata\n",
    "positive_control = \"c2\"  # @param {type:\"string\"}  # (str) Label for positive control\n",
    "negative_control = \"c1\"  # @param {type:\"string\"}  # (str) Label for negative control\n",
    "exclude = None  # @param {type:\"raw\"}  # (str or None) Rows to exclude in location_column\n",
    "\n",
    "# === Filtering and data quality ===\n",
    "nuclei_limit = 1  # @param {type:\"integer\"}  # (int) Max number of nuclei per cell\n",
    "pathogen_limit = 3  # @param {type:\"integer\"}  # (int) Max number of pathogens per cell\n",
    "minimum_cell_count = 25  # @param {type:\"integer\"}  # (int) Minimum number of cells per well\n",
    "\n",
    "# === Heatmap and features ===\n",
    "heatmap_feature = \"predictions\"  # @param {type:\"string\"}  # (str) Column to display in heatmap\n",
    "grouping = \"mean\"  # @param [\"mean\", \"median\"] {type:\"string\"}  # (str) Grouping method\n",
    "min_max = \"allq\"  # @param [\"all\", \"allq\"] {type:\"string\"}  # (str) Quantile normalization\n",
    "cmap = \"viridis\"  # @param {type:\"string\"}  # (str) Colormap for heatmap\n",
    "\n",
    "# === Feature selection and importance ===\n",
    "remove_low_variance_features = True  # @param {type:\"boolean\"}  # (bool) Remove low-variance features\n",
    "remove_highly_correlated_features = True  # @param {type:\"boolean\"}  # (bool) Remove highly correlated features\n",
    "n_repeats = 10  # @param {type:\"integer\"}  # (int) Repeats for permutation importance\n",
    "top_features = 30  # @param {type:\"integer\"}  # (int) Number of top features to display\n",
    "channel_of_interest = 1  # @param {type:\"integer\"}  # (int) Channel used for class-specific analysis\n",
    "\n",
    "# === Runtime ===\n",
    "verbose = False  # @param {type:\"boolean\"}  # (bool) Display verbose output\n",
    "n_jobs = 10  # @param {type:\"integer\"}  # (int) Number of threads to use\n",
    "\n",
    "# === Assemble dictionary and run ===\n",
    "settings = {\n",
    "    'src': src,\n",
    "    'model_type_ml': model_type_ml,\n",
    "    'heatmap_feature': heatmap_feature,\n",
    "    'grouping': grouping,\n",
    "    'min_max': min_max,\n",
    "    'cmap': cmap,\n",
    "    'n_estimators': n_estimators,\n",
    "    'test_size': test_size,\n",
    "    'location_column': location_column,\n",
    "    'positive_control': positive_control,\n",
    "    'negative_control': negative_control,\n",
    "    'exclude': exclude,\n",
    "    'nuclei_limit': nuclei_limit,\n",
    "    'pathogen_limit': pathogen_limit,\n",
    "    'n_repeats': n_repeats,\n",
    "    'top_features': top_features,\n",
    "    'channel_of_interest': channel_of_interest,\n",
    "    'minimum_cell_count': minimum_cell_count,\n",
    "    'remove_low_variance_features': remove_low_variance_features,\n",
    "    'remove_highly_correlated_features': remove_highly_correlated_features,\n",
    "    'verbose': verbose,\n",
    "    'n_jobs': n_jobs\n",
    "}\n",
    "\n",
    "results = generate_ml_scores(settings)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e4dde9e-2046-4327-975e-d08dd7455a43",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.ml import perform_regression\n",
    "import pandas as pd\n",
    "%matplotlib inline\n",
    "\n",
    "# @title Regression Settings { display-mode: \"form\" }\n",
    "\n",
    "# === Input Data ===\n",
    "count_data = \"path\"  # @param {type:\"string\"}  # (path) Path or list of paths to sequencing count data\n",
    "score_data = \"path\"  # @param {type:\"string\"}  # (path) Path or list of paths to score data\n",
    "score_column = \"column\"  # @param {type:\"string\"}  # (str) Column with cell scores\n",
    "metadata_files = [\"path.csv\", \"path.csv\"]  # @param {type:\"raw\"}  # (list) Paths to gene metadata CSV files\n",
    "\n",
    "# === Control Genes and Thresholds ===\n",
    "positive_control = \"gene\"  # @param {type:\"string\"}  # (str) Gene to highlight in volcano plot\n",
    "negative_control = \"gene\"  # @param {type:\"string\"}  # (str) Gene to highlight in volcano plot\n",
    "min_n = 3  # @param {type:\"integer\"}  # (int) Minimum number of cells per gRNA\n",
    "fraction_threshold = None  # @param {type:\"raw\"}  # (float or None) Minimum gene fraction threshold\n",
    "\n",
    "# === gRNA/Well Filtering ===\n",
    "target_unique_count = 5  # @param {type:\"integer\"}  # (int) Number of expected unique gRNAs per well\n",
    "tolerance = 0.02  # @param {type:\"number\"}  # (float) Tolerance for cell-per-well limit\n",
    "min_cell_count = None  # @param {type:\"raw\"}  # (int or None) Minimum number of cells per well\n",
    "control_wells = [\"c1\", \"c2\", \"c3\"]  # @param {type:\"raw\"}  # (list) Metadata to exclude from regression model\n",
    "filter_column = \"column\"  # @param {type:\"string\"}  # (str) Column containing control metadata to remove\n",
    "\n",
    "# === Regression Parameters ===\n",
    "dependent_variable = \"column\"  # @param {type:\"string\"}  # (str) Column to regress\n",
    "regression_type = \"ols\"  # @param [\"ols\", \"glm\", \"mixed\", \"ridge\", \"lasso\"] {type:\"string\"}  # (str) Type of regression\n",
    "random_row_column_effects = False  # @param {type:\"boolean\"}  # (bool) Include row/column/plate effects\n",
    "cov_type = None  # @param {type:\"raw\"}  # (str or None) Covariance estimator for OLS\n",
    "plate = None  # @param {type:\"raw\"}  # (str or None) Replace plate values\n",
    "\n",
    "# === Output Configuration ===\n",
    "threshold_method = \"var\"  # @param [\"std\", \"var\"] {type:\"string\"}  # (str) Thresholding method\n",
    "threshold_multiplier = 4  # @param {type:\"integer\"}  # (int) Multiplier for effect size threshold\n",
    "transform = \"log\"  # @param {type:\"string\"}  # (str) Transformation for dependent variable\n",
    "agg_type = \"mean\"  # @param [\"mean\", \"median\"] {type:\"string\"}  # (str) Aggregation for dependent variable\n",
    "volcano = \"gene\"  # @param [\"gene\", \"grna\", \"all\"] {type:\"string\"}  # (str) Volcano result display mode\n",
    "alpha = 0.8  # @param {type:\"number\"}  # (float) Alpha for hinge/lasso regressions\n",
    "log_x = False  # @param {type:\"boolean\"}  # (bool) Log scale for X axis\n",
    "log_y = False  # @param {type:\"boolean\"}  # (bool) Log scale for Y axis\n",
    "y_lims = [[0, 9], [12, 16]]  # @param {type:\"raw\"}  # (list of lists) Limits for broken Y axis\n",
    "\n",
    "# === Execute Regression ===\n",
    "settings = {\n",
    "    'count_data': count_data,\n",
    "    'score_data': score_data,\n",
    "    'score_column': score_column,\n",
    "    'metadata_files': metadata_files,\n",
    "    'positive_control': positive_control,\n",
    "    'negative_control': negative_control,\n",
    "    'min_n': min_n,\n",
    "    'fraction_threshold': fraction_threshold,\n",
    "    'target_unique_count': target_unique_count,\n",
    "    'tolerance': tolerance,\n",
    "    'log_x': log_x,\n",
    "    'log_y': log_y,\n",
    "    'x_lim': None,\n",
    "    'control_wells': control_wells,\n",
    "    'filter_column': filter_column,\n",
    "    'dependent_variable': dependent_variable,\n",
    "    'threshold_method': threshold_method,\n",
    "    'threshold_multiplier': threshold_multiplier,\n",
    "    'transform': transform,\n",
    "    'agg_type': agg_type,\n",
    "    'min_cell_count': min_cell_count,\n",
    "    'regression_type': regression_type,\n",
    "    'random_row_column_effects': random_row_column_effects,\n",
    "    'y_lims': y_lims,\n",
    "    'plate': plate,\n",
    "    'cov_type': cov_type,\n",
    "    'volcano': volcano,\n",
    "    'alpha': alpha\n",
    "}\n",
    "\n",
    "coef_df = perform_regression(settings)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b7a8b5b-93f2-4b5c-884d-b4cc62fa0c71",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.io import generate_training_dataset\n",
    "%matplotlib inline\n",
    "\n",
    "# @title Training Dataset Settings { display-mode: \"form\" }\n",
    "\n",
    "src = \"path\"  # @param {type:\"string\"}  # (path) Path to source folder (where original images are stored)\n",
    "dataset_mode = \"metadata_annotation\"  # @param [\"annotation\", \"measurement\", \"metadata\", \"annotation_metadata\"] {type:\"string\"}  # (string) Mode for dataset generation\n",
    "tables = [\"cell\"]  # @param {type:\"raw\"}  # (list of strings) Tables present in the database (excluding png_list)\n",
    "test_split = 0.1  # @param {type:\"number\"}  # (float) Fraction of images used for test set\n",
    "\n",
    "annotation_column = \"test\"  # @param {type:\"string\"}  # (Optional, string) Column used for annotation labels\n",
    "annotated_classes = [1]  # @param {type:\"raw\"}  # (Optional, list of integers) Class labels in annotation column\n",
    "\n",
    "metadata_type_by = \"column_name\"  # @param {type:\"string\"}  # (Optional, string) Column that defines metadata classes\n",
    "class_metadata = [\"c10\", \"c11\", \"c12\", \"c22\", \"c23\", \"c24\"]  # @param {type:\"raw\"}  # (Optional, list) Metadata values defining classes\n",
    "\n",
    "png_type = \"cell_png\"  # @param {type:\"string\"}  # (Optional, string) Filter by image path substring\n",
    "nuclei_limit = False  # @param {type:\"boolean\"}  # (Optional, bool) Filter for number of nuclei per cell\n",
    "pathogen_limit = 0  # @param {type:\"integer\"}  # (Optional, int) Filter for number of pathogens per cell\n",
    "uninfected = True  # @param {type:\"boolean\"}  # (Optional, bool) Include uninfected cells\n",
    "size = None  # @param {type:\"raw\"}  # (Optional, int or None) Max number of images per class\n",
    "\n",
    "settings = {\n",
    "    'src': src,\n",
    "    'dataset_mode': dataset_mode,\n",
    "    'tables': tables,\n",
    "    'test_split': test_split,\n",
    "    'annotation_column': annotation_column,\n",
    "    'annotated_classes': annotated_classes,\n",
    "    'metadata_type_by': metadata_type_by,\n",
    "    'class_metadata': class_metadata,\n",
    "    'png_type': png_type,\n",
    "    'nuclei_limit': nuclei_limit,\n",
    "    'pathogen_limit': pathogen_limitHow to Cite\n",
    "-----------\n",
    "If you use spaCR in your research, please cite:  \n",
    "Olafsson EB, et al. SpaCr: Spatial phenotype analysis of CRISPR-Cas9 screens. *Manuscript in preparation*.,\n",
    "    'uninfected': uninfected,\n",
    "    'size': size\n",
    "}\n",
    "\n",
    "generate_training_dataset(settings)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c9130db-61b5-4d0c-8c47-74c2834f298b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.deep_spacr import train_test_model\n",
    "%matplotlib inline\n",
    "\n",
    "# @title Deep Learning Model Training Settings { display-mode: \"form\" }\n",
    "\n",
    "src = \"path\"  # @param {type:\"string\"}  # (path) Path to source folder (ends with datasets/training)\n",
    "train = False  # @param {type:\"boolean\"}  # (bool) Train the model\n",
    "test = True  # @param {type:\"boolean\"}  # (bool) Test the model\n",
    "custom_model = \"path\"  # @param {type:\"string\"}  # (path) Path to a custom model\n",
    "\n",
    "classes = [\"nc\", \"pc\"]  # @param {type:\"raw\"}  # (list) Class folder names in datasets/training/train or test\n",
    "model_type = \"maxvit_t\"  # @param {type:\"string\"}  # (string) Torch model architecture\n",
    "optimizer_type = \"adamw\"  # @param {type:\"string\"}  # (string) Optimizer type\n",
    "schedule = \"reduce_lr_on_plateau\"  # @param [\"reduce_lr_on_plateau\", \"step_lr\"] {type:\"string\"}  # (string) Scheduler type\n",
    "loss_type = \"focal_loss\"  # @param [\"focal_loss\", \"binary_cross_entropy_with_logits\"] {type:\"string\"}  # (string) Loss function\n",
    "\n",
    "normalize = True  # @param {type:\"boolean\"}  # (bool) Apply ImageNet normalization\n",
    "image_size = 224  # @param {type:\"integer\"}  # (int) Size of input images\n",
    "batch_size = 64  # @param {type:\"integer\"}  # (int) Number of images per batch\n",
    "epochs = 100  # @param {type:\"integer\"}  # (int) Number of training epochs\n",
    "val_split = 0.1  # @param {type:\"number\"}  # (float) Fraction of images for validation\n",
    "\n",
    "learning_rate = 0.0001  # @param {type:\"number\"}  # (float) Learning rate\n",
    "weight_decay = 0.00001  # @param {type:\"number\"}  # (float) Weight decay for regularization\n",
    "dropout_rate = 0.1  # @param {type:\"number\"}  # (float) Dropout rate\n",
    "init_weights = True  # @param {type:\"boolean\"}  # (bool) Use pretrained ImageNet weights\n",
    "\n",
    "amsgrad = True  # @param {type:\"boolean\"}  # (bool) Use AMSGrad optimizer variant\n",
    "use_checkpoint = True  # @param {type:\"boolean\"}  # (bool) Use gradient checkpointing to save VRAM\n",
    "gradient_accumulation = True  # @param {type:\"boolean\"}  # (bool) Accumulate gradients over steps\n",
    "gradient_accumulation_steps = 4  # @param {type:\"integer\"}  # (int) Number of steps to accumulate gradients\n",
    "\n",
    "intermedeate_save = True  # @param {type:\"boolean\"}  # (bool) Save intermediate model states\n",
    "pin_memory = True  # @param {type:\"boolean\"}  # (bool) Enable pinning memory in DataLoader\n",
    "n_jobs = 30  # @param {type:\"integer\"}  # (int) Number of parallel jobs (threads)\n",
    "train_channels = [\"r\", \"g\", \"b\"]  # @param {type:\"raw\"}  # (list of 'r', 'g', 'b') Channels used in training\n",
    "augment = False  # @param {type:\"boolean\"}  # (bool) Enable dataset augmentation\n",
    "verbose = True  # @param {type:\"boolean\"}  # (bool) Verbose output\n",
    "\n",
    "settings = {\n",
    "    'src': src,\n",
    "    'train': train,\n",
    "    'test': test,\n",
    "    'custom_model': custom_model,\n",
    "    'classes': classes,\n",
    "    'model_type': model_type,\n",
    "    'optimizer_type': optimizer_type,\n",
    "    'schedule': schedule,\n",
    "    'loss_type': loss_type,\n",
    "    'normalize': normalize,\n",
    "    'image_size': image_size,\n",
    "    'batch_size': batch_size,\n",
    "    'epochs': epochs,\n",
    "    'val_split': val_split,\n",
    "    'learning_rate': learning_rate,\n",
    "    'weight_decay': weight_decay,\n",
    "    'dropout_rate': dropout_rate,\n",
    "    'init_weights': init_weights,\n",
    "    'amsgrad': amsgrad,\n",
    "    'use_checkpoint': use_checkpoint,\n",
    "    'gradient_accumulation': gradient_accumulation,\n",
    "    'gradient_accumulation_steps': gradient_accumulation_steps,\n",
    "    'intermedeate_save': intermedeate_save,\n",
    "    'pin_memory': pin_memory,\n",
    "    'n_jobs': n_jobs,\n",
    "    'train_channels': train_channels,\n",
    "    'augment': augment,\n",
    "    'verbose': verbose\n",
    "}\n",
    "\n",
    "train_test_model(settings)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e899cd7-eb9e-49ad-a612-5ae95283fef5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.io import generate_dataset\n",
    "%matplotlib inline\n",
    "\n",
    "# @title Generate TAR Dataset Settings { display-mode: \"form\" }\n",
    "\n",
    "src = \"path\"  # @param {type:\"string\"}  # (path) Path to source folder (where original images were stored)\n",
    "file_metadata = \"cell_png\"  # @param {type:\"string\"}  # (Optional, string) Substring in file path used to filter images\n",
    "experiment = \"test\"  # @param {type:\"string\"}  # (string) Name of dataset\n",
    "sample = 10000  # @param {type:\"integer\"}  # (Optional, integer) Maximum number of images to include\n",
    "\n",
    "settings = {\n",
    "    'src': src,\n",
    "    'file_metadata': file_metadata,\n",
    "    'experiment': experiment,\n",
    "    'sample': sample\n",
    "}\n",
    "\n",
    "generate_dataset(settings)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a71df0a-9bba-48ad-a45d-4b3df6daae95",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.deep_spacr import apply_model_to_tar\n",
    "%matplotlib inline\n",
    "\n",
    "# @title Apply Model to TAR Dataset Settings { display-mode: \"form\" }\n",
    "\n",
    "dataset = \"path.tar\"  # @param {type:\"string\"}  # (path) Path to TAR dataset (ends with .tar)\n",
    "model_path = \"path.pth\"  # @param {type:\"string\"}  # (path) Path to model file (ends with .pth)\n",
    "file_type = \"cell_png\"  # @param {type:\"string\"}  # (Optional, string) Filter images by substring in file path\n",
    "image_size = 224  # @param {type:\"integer\"}  # (int) Size of input images (height and width)\n",
    "batch_size = 64  # @param {type:\"integer\"}  # (int) Number of images per batch\n",
    "normalize = True  # @param {type:\"boolean\"}  # (bool) Apply ImageNet normalization\n",
    "score_threshold = 0.5  # @param {type:\"number\"}  # (float) Threshold for classification score\n",
    "n_jobs = 30  # @param {type:\"integer\"}  # (int) Number of threads to use\n",
    "verbose = True  # @param {type:\"boolean\"}  # (bool) Enable verbose output\n",
    "\n",
    "settings = {\n",
    "    'dataset': dataset,\n",
    "    'model_path': model_path,\n",
    "    'file_type': file_type,\n",
    "    'image_size': image_size,\n",
    "    'batch_size': batch_size,\n",
    "    'normalize': normalize,\n",
    "    'score_threshold': score_threshold,\n",
    "    'n_jobs': n_jobs,\n",
    "    'verbose': verbose\n",
    "}\n",
    "\n",
    "result_df = apply_model_to_tar(settings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ac9ba41-3a84-45a4-b7f5-ab5845fdbdd3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f0fb45d-805e-4059-a4b1-29eb29483a9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from spacr.sequencing import generate_barecode_mapping\n",
    "%matplotlib inline\n",
    "\n",
    "# @title Generate Barcode Mapping Settings { display-mode: \"form\" }\n",
    "\n",
    "src = \"path\"  # @param {type:\"string\"}  # (path) Path to source folder\n",
    "regex = \"^(?P<column>.{8})TGCTG.*TAAAC(?P<grna>.{20,21})AACTT.*AGAAG(?P<row_name>.{8}).*\"  # @param {type:\"string\"}  # (regex) Pattern to extract barcodes\n",
    "target_sequence = \"TGCTGTTTCCAGCATAGCTCTTAAAC\"  # @param {type:\"string\"}  # (string) Target sequence to locate grna\n",
    "offset_start = -8  # @param {type:\"integer\"}  # (int) Offset for grna matching\n",
    "expected_end = 89  # @param {type:\"integer\"}  # (int) Expected final position\n",
    "column_csv = \"path to column_barecodes.csv\"  # @param {type:\"string\"}  # (path) CSV file with column barcodes\n",
    "grna_csv = \"path to grna_barcodes_RC.csv\"  # @param {type:\"string\"}  # (path) CSV file with grna barcodes\n",
    "row_csv = \"path to row_barecodes_RC.csv\"  # @param {type:\"string\"}  # (path) CSV file with row barcodes\n",
    "save_h5 = True  # @param {type:\"boolean\"}  # (bool) Save results to HDF5\n",
    "comp_type = \"zlib\"  # @param [\"zlib\", \"gzip\", \"lzf\", null] {type:\"string\"}  # (string) Compression type\n",
    "comp_level = 5  # @param {type:\"integer\"}  # (int) Compression level\n",
    "chunk_size = 10000  # @param {type:\"integer\"}  # (int) Number of reads per chunk\n",
    "n_jobs = None  # @param {type:\"raw\"}  # (int or None) Number of parallel jobs\n",
    "mode = \"paired\"  # @param [\"paired\", \"single\"] {type:\"string\"}  # (string) Read mode\n",
    "single_direction = \"R1\"  # @param [\"R1\", \"R2\"] {type:\"string\"}  # (string) Direction if mode is single\n",
    "test = False  # @param {type:\"boolean\"}  # (bool) Run in test mode\n",
    "fill_na = True  # @param {type:\"boolean\"}  # (bool) Fill missing data with NA\n",
    "\n",
    "settings = {\n",
    "    'src': src,\n",
    "    'regex': regex,\n",
    "    'target_sequence': target_sequence,\n",
    "    'offset_start': offset_start,\n",
    "    'expected_end': expected_end,\n",
    "    'column_csv': column_csv,\n",
    "    'grna_csv': grna_csv,\n",
    "    'row_csv': row_csv,\n",
    "    'save_h5': save_h5,\n",
    "    'comp_type': comp_type,\n",
    "    'comp_level': comp_level,\n",
    "    'chunk_size': chunk_size,\n",
    "    'n_jobs': n_jobs,\n",
    "    'mode': mode,\n",
    "    'single_direction': single_direction,\n",
    "    'test': test,\n",
    "    'fill_na': fill_na\n",
    "}\n",
    "\n",
    "generate_barecode_mapping(settings)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab28cd99-0395-4137-8b63-f3fe56611035",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Description: Prepare a dataset for Cellpose training\n",
    "from spacr.io import prepare_cellpose_dataset\n",
    "%matplotlib inline\n",
    "\n",
    "# @title Prepare Cellpose Dataset Settings { display-mode: \"form\" }\n",
    "\n",
    "input_root = \"path\"  # @param {type:\"string\"}  # (path) Path to the dataset root folder\n",
    "augment_data = True  # @param {type:\"boolean\"}  # (bool) Augment dataset by flipping and rotating images\n",
    "train_fraction = 0.8  # @param {type:\"number\"}  # (float) Fraction of data to use for training\n",
    "n_jobs = None  # @param {type:\"raw\"}  # (int or None) Number of parallel jobs (None = auto)\n",
    "\n",
    "prepare_cellpose_dataset(\n",
    "    input_root=input_root,\n",
    "    augment_data=augment_data,\n",
    "    train_fraction=train_fraction,\n",
    "    n_jobs=n_jobs\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be26e4e1-6ec4-4d44-bcc3-70d5f942d44e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Description: Train a Cellpose model\n",
    "from spacr.submodules import train_cellpose\n",
    "%matplotlib inline\n",
    "\n",
    "# @title Train Cellpose Settings { display-mode: \"form\" }\n",
    "\n",
    "src = \"path\"  # @param {type:\"string\"}  # (path) path to training dataset\n",
    "test = False  # @param {type:\"boolean\"}  # (bool) Run in test mode\n",
    "normalize = False  # @param {type:\"boolean\"}  # (bool) Normalize images\n",
    "percentiles = None  # @param {type:\"raw\"}  # (tuple or None) Percentiles to normalize to\n",
    "invert = False  # @param {type:\"boolean\"}  # (bool) Invert image intensities\n",
    "grayscale = True  # @param {type:\"boolean\"}  # (bool) Convert to grayscale\n",
    "rescale = False  # @param {type:\"boolean\"}  # (bool) Rescale intensities\n",
    "circular = False  # @param {type:\"boolean\"}  # (bool) Assume circular objects\n",
    "channels = [0, 0]  # @param {type:\"raw\"}  # (list) Channels used for segmentation\n",
    "model_name = \"test\"  # @param {type:\"string\"}  # (string) Name for saving model\n",
    "model_type = \"cyto\"  # @param [\"cyto\", \"nuclei\", \"cyto2\"] {type:\"string\"}  # (string) Type of Cellpose model\n",
    "Signal_to_noise = 10  # @param {type:\"integer\"}  # (int) Signal to noise ratio\n",
    "background = 200  # @param {type:\"integer\"}  # (int) Background intensity\n",
    "remove_background = False  # @param {type:\"boolean\"}  # (bool) Remove background\n",
    "learning_rate = 0.2  # @param {type:\"number\"}  # (float) Learning rate\n",
    "weight_decay = 1e-05  # @param {type:\"number\"}  # (float) Weight decay\n",
    "batch_size = 8  # @param {type:\"integer\"}  # (int) Batch size\n",
    "n_epochs = 25000  # @param {type:\"integer\"}  # (int) Number of training epochs\n",
    "from_scratch = False  # @param {type:\"boolean\"}  # (bool) Train from scratch\n",
    "diameter = 30  # @param {type:\"integer\"}  # (int) Expected diameter of objects\n",
    "resize = False  # @param {type:\"boolean\"}  # (bool) Resize images before training\n",
    "target_dimensions = 1000  # @param {type:\"integer\"}  # (int) Target image dimension after resizing\n",
    "verbose = True  # @param {type:\"boolean\"}  # (bool) Verbose output\n",
    "\n",
    "settings = {\n",
    "    'src': src,\n",
    "    'test': test,\n",
    "    'normalize': normalize,\n",
    "    'percentiles': percentiles,\n",
    "    'invert': invert,\n",
    "    'grayscale': grayscale,\n",
    "    'rescale': rescale,\n",
    "    'circular': circular,\n",
    "    'channels': channels,\n",
    "    'model_name': model_name,\n",
    "    'model_type': model_type,\n",
    "    'Signal_to_noise': Signal_to_noise,\n",
    "    'background': background,\n",
    "    'remove_background': remove_background,\n",
    "    'learning_rate': learning_rate,\n",
    "    'weight_decay': weight_decay,\n",
    "    'batch_size': batch_size,\n",
    "    'n_epochs': n_epochs,\n",
    "    'from_scratch': from_scratch,\n",
    "    'diameter': diameter,\n",
    "    'resize': resize,\n",
    "    'target_dimensions': target_dimensions,\n",
    "    'verbose': verbose\n",
    "}\n",
    "\n",
    "train_cellpose(settings)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "223b0906-b8ab-47e6-90a0-112aeef8f011",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "spacr",
   "language": "python",
   "name": "spacr"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
