

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>spacr.ml &mdash; spacr 1.0.8 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../../_static/graphviz.css?v=fd3f3429" />
      <link rel="stylesheet" type="text/css" href="../../../_static/custom.css?v=45a07ac9" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=aec50437"></script>
      <script src="../../../_static/doctools.js?v=9a2dae69"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="spacr.plot" href="../plot/index.html" />
    <link rel="prev" title="spacr.measure" href="../measure/index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: #005f73" >

          
          
          <a href="../../../index.html">
            
              <img src="../../../_static/logo_spacr.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../../index.html">API Reference</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="../index.html">spacr</a><ul class="current">
<li class="toctree-l3 current"><a class="reference internal" href="../index.html#submodules">Submodules</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="../app_classify/index.html">spacr.app_classify</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app_mask/index.html">spacr.app_mask</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app_measure/index.html">spacr.app_measure</a></li>
<li class="toctree-l4"><a class="reference internal" href="../app_sequencing/index.html">spacr.app_sequencing</a></li>
<li class="toctree-l4"><a class="reference internal" href="../core/index.html">spacr.core</a></li>
<li class="toctree-l4"><a class="reference internal" href="../deep_spacr/index.html">spacr.deep_spacr</a></li>
<li class="toctree-l4"><a class="reference internal" href="../gui/index.html">spacr.gui</a></li>
<li class="toctree-l4"><a class="reference internal" href="../gui_core/index.html">spacr.gui_core</a></li>
<li class="toctree-l4"><a class="reference internal" href="../gui_elements/index.html">spacr.gui_elements</a></li>
<li class="toctree-l4"><a class="reference internal" href="../gui_utils/index.html">spacr.gui_utils</a></li>
<li class="toctree-l4"><a class="reference internal" href="../io/index.html">spacr.io</a></li>
<li class="toctree-l4"><a class="reference internal" href="../logger/index.html">spacr.logger</a></li>
<li class="toctree-l4"><a class="reference internal" href="../measure/index.html">spacr.measure</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">spacr.ml</a></li>
<li class="toctree-l4"><a class="reference internal" href="../plot/index.html">spacr.plot</a></li>
<li class="toctree-l4"><a class="reference internal" href="../sequencing/index.html">spacr.sequencing</a></li>
<li class="toctree-l4"><a class="reference internal" href="../settings/index.html">spacr.settings</a></li>
<li class="toctree-l4"><a class="reference internal" href="../sp_stats/index.html">spacr.sp_stats</a></li>
<li class="toctree-l4"><a class="reference internal" href="../spacr_cellpose/index.html">spacr.spacr_cellpose</a></li>
<li class="toctree-l4"><a class="reference internal" href="../submodules/index.html">spacr.submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="../timelapse/index.html">spacr.timelapse</a></li>
<li class="toctree-l4"><a class="reference internal" href="../toxo/index.html">spacr.toxo</a></li>
<li class="toctree-l4"><a class="reference internal" href="../utils/index.html">spacr.utils</a></li>
<li class="toctree-l4"><a class="reference internal" href="../version/index.html">spacr.version</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu"  style="background: #005f73" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">spacr</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">API Reference</a></li>
          <li class="breadcrumb-item"><a href="../index.html">spacr</a></li>
      <li class="breadcrumb-item active">spacr.ml</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/api/spacr/ml/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-spacr.ml">
<span id="spacr-ml"></span><h1>spacr.ml<a class="headerlink" href="#module-spacr.ml" title="Link to this heading"></a></h1>
<section id="module-contents">
<h2>Module Contents<a class="headerlink" href="#module-contents" title="Link to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="spacr.ml.QuasiBinomial">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">QuasiBinomial</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">link</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">logit()</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dispersion</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#QuasiBinomial"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.QuasiBinomial" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-obj docutils literal notranslate"><span class="pre">statsmodels.genmod.families.Binomial</span></code></p>
<p>Custom Quasi-Binomial family with adjustable variance.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="spacr.ml.QuasiBinomial.dispersion">
<span class="sig-name descname"><span class="pre">dispersion</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">1.0</span></em><a class="reference internal" href="../../../_modules/spacr/ml.html#QuasiBinomial.dispersion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.QuasiBinomial.dispersion" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="spacr.ml.QuasiBinomial.variance">
<span class="sig-name descname"><span class="pre">variance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mu</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#QuasiBinomial.variance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.QuasiBinomial.variance" title="Link to this definition"></a></dt>
<dd><p>Adjust the variance with the dispersion parameter.</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.calculate_p_values">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">calculate_p_values</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#calculate_p_values"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.calculate_p_values" title="Link to this definition"></a></dt>
<dd><p>Calculate p-values for model coefficients using residuals and t-statistics.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>np.ndarray</em>) – The input feature matrix used for fitting the model.</p></li>
<li><p><strong>y</strong> (<em>np.ndarray</em>) – The target variable.</p></li>
<li><p><strong>model</strong> (<em>object</em>) – A fitted sklearn linear model (must have <cite>.predict</cite> and <cite>.coef_</cite> attributes).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An array of p-values corresponding to each coefficient in the model.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.perform_mixed_model">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">perform_mixed_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#perform_mixed_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.perform_mixed_model" title="Link to this definition"></a></dt>
<dd><p>Perform mixed effects regression and return the fitted model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>y</strong> (<em>pd.Series</em><em> or </em><em>np.ndarray</em>) – Target variable.</p></li>
<li><p><strong>X</strong> (<em>pd.DataFrame</em>) – Feature matrix.</p></li>
<li><p><strong>groups</strong> (<em>pd.Series</em><em> or </em><em>np.ndarray</em>) – Grouping variable for mixed model.</p></li>
<li><p><strong>alpha</strong> (<em>float</em><em>, </em><em>optional</em>) – Regularization strength for Ridge regression when multicollinearity is detected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Fitted mixed model results.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>statsmodels.regression.mixed_linear_model.MixedLMResults</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.create_volcano_filename">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">create_volcano_filename</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">csv_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regression_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dst</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#create_volcano_filename"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.create_volcano_filename" title="Link to this definition"></a></dt>
<dd><p>Construct the file name for saving the volcano plot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>csv_path</strong> (<em>str</em>) – Path to the CSV file containing model results.</p></li>
<li><p><strong>regression_type</strong> (<em>str</em>) – Type of regression performed.</p></li>
<li><p><strong>alpha</strong> (<em>float</em>) – Alpha value (used for quantile regression naming).</p></li>
<li><p><strong>dst</strong> (<em>str</em><em> or </em><em>None</em>) – Destination directory to save the file.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Full path for the output volcano plot PDF file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.scale_variables">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">scale_variables</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#scale_variables"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.scale_variables" title="Link to this definition"></a></dt>
<dd><p>Min-max scale both independent and dependent variables.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>pd.DataFrame</em>) – Feature matrix.</p></li>
<li><p><strong>y</strong> (<em>np.ndarray</em>) – Target variable.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>X_scaled</strong> (<em>pd.DataFrame</em>) – Scaled feature matrix.</p></li>
<li><p><strong>y_scaled</strong> (<em>np.ndarray</em>) – Scaled target variable.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.select_glm_family">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">select_glm_family</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">y</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#select_glm_family"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.select_glm_family" title="Link to this definition"></a></dt>
<dd><p>Automatically select a GLM family based on the nature of the dependent variable.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>y</strong> (<em>np.ndarray</em>) – Target variable.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>GLM family appropriate for the data.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>sm.families.Family</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.prepare_formula">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">prepare_formula</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dependent_variable</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_row_column_effects</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#prepare_formula"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.prepare_formula" title="Link to this definition"></a></dt>
<dd><p>Prepare the formula for regression modeling based on model design.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dependent_variable</strong> (<em>str</em>) – Name of the dependent variable.</p></li>
<li><p><strong>random_row_column_effects</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to include row and column IDs as random effects.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Regression formula string.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.fit_mixed_model">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">fit_mixed_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">formula</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dst</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#fit_mixed_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.fit_mixed_model" title="Link to this definition"></a></dt>
<dd><p>Fit a mixed linear model with random effects for plate, row, and column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – Input data.</p></li>
<li><p><strong>formula</strong> (<em>str</em>) – Regression formula.</p></li>
<li><p><strong>dst</strong> (<em>str</em>) – Output folder to save diagnostic plots.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>mixed_model</strong> (<em>statsmodels.regression.mixed_linear_model.MixedLMResults</em>) – Fitted model.</p></li>
<li><p><strong>coef_df</strong> (<em>pd.DataFrame</em>) – DataFrame of coefficients and p-values.</p></li>
</ul>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.check_and_clean_data">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">check_and_clean_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependent_variable</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#check_and_clean_data"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.check_and_clean_data" title="Link to this definition"></a></dt>
<dd><p>Validate and preprocess a DataFrame before model fitting.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – Input data.</p></li>
<li><p><strong>dependent_variable</strong> (<em>str</em>) – Name of the dependent variable column.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Cleaned and validated DataFrame, ready for model fitting.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.minimum_cell_simulation">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">minimum_cell_simulation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">settings</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">num_repeats</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sample_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.02</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">increment</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#minimum_cell_simulation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.minimum_cell_simulation" title="Link to this definition"></a></dt>
<dd><p>Estimate the minimum number of cells required per well to stabilize phenotype measurements.</p>
<p>This function simulates phenotype score stability by repeatedly subsampling increasing numbers
of cells per well, calculating the mean absolute difference from the full-well mean. It identifies
the minimal sample size (elbow point) at which the average difference drops below a user-defined
tolerance, smoothed across wells.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>settings</strong> (<em>dict</em>) – A dictionary with keys:
- ‘score_data’ (list or str): CSV path(s) with cell-level data.
- ‘score_column’ (str): Column name of the phenotype score.
- ‘tolerance’ (float or int): Allowed deviation from true mean (e.g. 0.02 for 2%).
- ‘min_cell_count’ (int or None): Optional fixed value for annotation.
- ‘count_data’ (list): Used to define the output folder for figure saving.</p></li>
<li><p><strong>num_repeats</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of times to resample for each sample size. Default is 10.</p></li>
<li><p><strong>sample_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of top wells to simulate (by cell count). Default is 100.</p></li>
<li><p><strong>tolerance</strong> (<em>float</em><em> or </em><em>int</em><em>, </em><em>optional</em>) – Tolerance threshold to define the elbow point. Default is 0.02.</p></li>
<li><p><strong>smoothing</strong> (<em>int</em><em>, </em><em>optional</em>) – Window size for smoothing the mean absolute difference curve. Default is 10.</p></li>
<li><p><strong>increment</strong> (<em>int</em><em>, </em><em>optional</em>) – Step size between tested sample sizes. Default is 10.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Estimated minimal required number of cells (elbow point).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.process_model_coefficients">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">process_model_coefficients</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regression_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">controls</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#process_model_coefficients"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.process_model_coefficients" title="Link to this definition"></a></dt>
<dd><p>Extract model coefficients, standard errors, and p-values into a DataFrame with annotations.</p>
<p>Supports various regression types including beta, OLS, GLM, ridge, lasso, and quasi-binomial.
Adds classification labels (nc, pc, control, other) and computes -log10(p-values).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – Fitted regression model object.</p></li>
<li><p><strong>regression_type</strong> (<em>str</em>) – Type of regression (e.g., ‘beta’, ‘ols’, ‘glm’, ‘ridge’, ‘lasso’).</p></li>
<li><p><strong>X</strong> (<em>pd.DataFrame</em>) – Feature matrix used for fitting the model.</p></li>
<li><p><strong>y</strong> (<em>np.ndarray</em>) – Target values used in model fitting.</p></li>
<li><p><strong>nc</strong> (<em>str</em>) – Identifier for negative control features.</p></li>
<li><p><strong>pc</strong> (<em>str</em>) – Identifier for positive control features.</p></li>
<li><p><strong>controls</strong> (<em>list</em>) – List of gRNAs used as general controls.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Table of coefficients, p-values, and annotations.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.check_distribution">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">check_distribution</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">epsilon</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-06</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#check_distribution"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.check_distribution" title="Link to this definition"></a></dt>
<dd><p>Analyze distribution of the target variable and recommend appropriate regression model.</p>
<p>Checks for:
- Binary data (logit)
- Continuous [0, 1) data (beta)
- Continuous [0, 1] data (quasi-binomial)
- Normal distribution (OLS)
- Beta distribution fit
- Default to GLM otherwise</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>y</strong> (<em>np.ndarray</em>) – Dependent variable values.</p></li>
<li><p><strong>epsilon</strong> (<em>float</em><em>, </em><em>optional</em>) – Threshold for boundary proximity detection. Default is 1e-6.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Suggested regression model name (e.g., ‘logit’, ‘beta’, ‘ols’, ‘glm’).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.pick_glm_family_and_link">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">pick_glm_family_and_link</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">y</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#pick_glm_family_and_link"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.pick_glm_family_and_link" title="Link to this definition"></a></dt>
<dd><p>Select the appropriate GLM family and link function based on distribution of y.</p>
<p>Inspects binary, count, proportion, normal, and overdispersed data and maps to:
- Binomial with Logit link
- Gaussian with Identity link
- Poisson with Log link
- Inverse Gaussian with Log link
- Negative Binomial with Log link</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>y</strong> (<em>np.ndarray</em>) – Response variable to inspect.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>GLM family object with appropriate link function.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>sm.families.Family</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If data is suitable for Beta regression, which GLM cannot handle.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.regression_model">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">regression_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regression_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ols'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">groups</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cov_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#regression_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.regression_model" title="Link to this definition"></a></dt>
<dd><p>Fit a regression model of the specified type to the data.</p>
<p>Supports multiple regression types: OLS, GLM, beta, logit, probit, ridge, lasso, and mixed models.
Automatically performs hyperparameter tuning (alpha) for lasso and ridge if alpha is 0 or None.
For GLMs, prints McFadden’s pseudo R². For regularized models, prints MSE and coefficient table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>pd.DataFrame</em>) – Design matrix of independent variables.</p></li>
<li><p><strong>y</strong> (<em>pd.Series</em><em> or </em><em>np.ndarray</em>) – Dependent variable.</p></li>
<li><p><strong>regression_type</strong> (<em>str</em>) – Type of regression (‘ols’, ‘glm’, ‘beta’, ‘logit’, ‘probit’, ‘lasso’, ‘ridge’, ‘mixed’).</p></li>
<li><p><strong>groups</strong> (<em>array-like</em><em>, </em><em>optional</em>) – Grouping variable for mixed models.</p></li>
<li><p><strong>alpha</strong> (<em>float</em>) – Regularization strength for lasso and ridge.</p></li>
<li><p><strong>cov_type</strong> (<em>str</em><em>, </em><em>optional</em>) – Covariance estimator type for OLS (e.g., ‘HC3’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Fitted model object (statsmodels or sklearn estimator).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>model</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.regression">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">regression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">csv_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependent_variable</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'predictions'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regression_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_row_column_effects</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'233460'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'220950'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">controls</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dst</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cov_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#regression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.regression" title="Link to this definition"></a></dt>
<dd><p>Perform regression analysis on a DataFrame with optional plotting and mixed effects support.</p>
<p>Handles data cleaning, formula preparation, scaling, model fitting, and coefficient extraction.
Automatically determines appropriate regression type if not specified.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – Input data.</p></li>
<li><p><strong>csv_path</strong> (<em>str</em>) – Path to input CSV used for labeling plots/files.</p></li>
<li><p><strong>dependent_variable</strong> (<em>str</em>) – Name of column to regress against.</p></li>
<li><p><strong>regression_type</strong> (<em>str</em><em>, </em><em>optional</em>) – Type of regression (auto-detected if None).</p></li>
<li><p><strong>alpha</strong> (<em>float</em>) – Regularization parameter for lasso/ridge.</p></li>
<li><p><strong>random_row_column_effects</strong> (<em>bool</em>) – Whether to model row/column as random effects.</p></li>
<li><p><strong>nc</strong> (<em>str</em>) – Identifier for negative controls.</p></li>
<li><p><strong>pc</strong> (<em>str</em>) – Identifier for positive controls.</p></li>
<li><p><strong>controls</strong> (<em>list</em>) – List of general control gRNAs.</p></li>
<li><p><strong>dst</strong> (<em>str</em><em>, </em><em>optional</em>) – Destination folder for output files.</p></li>
<li><p><strong>cov_type</strong> (<em>str</em><em>, </em><em>optional</em>) – Covariance estimator for OLS.</p></li>
<li><p><strong>plot</strong> (<em>bool</em>) – Whether to generate volcano plot.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>(fitted model, coefficients DataFrame, regression type used)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.save_summary_to_file">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">save_summary_to_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'summary.csv'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#save_summary_to_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.save_summary_to_file" title="Link to this definition"></a></dt>
<dd><p>Save the model’s summary output to a text or CSV file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – Fitted statsmodels model with a <cite>.summary()</cite> method.</p></li>
<li><p><strong>file_path</strong> (<em>str</em>) – Output path to save the summary (e.g., ‘summary.txt’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.perform_regression">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">perform_regression</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">settings</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#perform_regression"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.perform_regression" title="Link to this definition"></a></dt>
<dd><p>Perform full regression analysis pipeline for pooled CRISPR-Cas9 screens.</p>
<p>This function integrates data loading, filtering, transformation, regression modeling,
statistical significance assessment, visualization, and optional downstream Toxoplasma-specific analysis.
It handles both grna-level and gene-level aggregation and outputs results with optional metadata integration.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>settings</strong> (<em>dict</em>) – Dictionary containing regression settings. Recommended to start with
<cite>get_perform_regression_default_settings()</cite> to initialize.</p>
</dd>
</dl>
<dl class="simple">
<dt>Required Keys in <cite>settings</cite>:</dt><dd><ul class="simple">
<li><p>score_data (str or list): Path(s) to CSV(s) containing phenotypic scores.</p></li>
<li><p>count_data (str or list): Path(s) to CSV(s) containing barcode abundance counts.</p></li>
<li><p>dependent_variable (str): Column name in <cite>score_data</cite> to use as dependent variable.</p></li>
<li><p>regression_type (str or None): Type of regression model to use (‘ols’, ‘glm’, ‘lasso’, etc.).</p></li>
<li><p>filter_value (list): Values to exclude from <cite>filter_column</cite>.</p></li>
<li><p>filter_column (str): Column in data to apply exclusion filter.</p></li>
<li><p>plateID (str): Identifier to use when reconstructing <cite>prc</cite> labels.</p></li>
<li><p>alpha (float): Regularization strength for lasso/ridge regressions.</p></li>
<li><p>random_row_column_effects (bool): If True, treat row/column as random effects.</p></li>
<li><p>negative_control (str): Label for negative controls.</p></li>
<li><p>positive_control (str): Label for positive controls.</p></li>
<li><p>controls (list): List of control gRNAs for estimating coefficient threshold.</p></li>
<li><p>volcano (str): Type of volcano plot to generate (‘all’, ‘gene’, ‘grna’).</p></li>
<li><p>transform (bool): Whether to transform dependent variable.</p></li>
<li><p>agg_type (str): Aggregation method for per-well phenotypes (‘mean’, ‘median’, etc.).</p></li>
<li><p>metadata_files (str or list): Metadata CSV(s) to merge with regression results.</p></li>
<li><p>toxo (bool): If True, generate Toxoplasma-specific downstream plots.</p></li>
<li><p>threshold_method (str): Method to calculate effect size threshold (‘std’ or ‘var’).</p></li>
<li><p>threshold_multiplier (float): Multiplier for threshold method.</p></li>
<li><p>outlier_detection (bool): Whether to remove grnas with outlier well counts.</p></li>
<li><p>cov_type (str or None): Covariance type for OLS (e.g., ‘HC3’).</p></li>
<li><p>verbose (bool): Whether to print verbose output.</p></li>
<li><p>x_lim, y_lims: Optional axes limits for volcano plots.</p></li>
<li><p>tolerance (float): Tolerance for determining minimal cell count via simulation.</p></li>
<li><p>min_cell_count (int or None): Minimum number of cells per well for inclusion.</p></li>
<li><p>min_n (int): Minimum number of replicates for inclusion in filtered results.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>‘results’: Full coefficient table (pandas DataFrame).
‘significant’: Filtered table of statistically significant hits (pandas DataFrame).</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<dl class="simple">
<dt>Side Effects:</dt><dd><ul class="simple">
<li><p>Saves regression results, metadata-merged results, volcano plots, and phenotype plots to disk.</p></li>
<li><p>Optionally generates publication-ready plots for Toxoplasma gene expression and phenotype scores.</p></li>
<li><p>Writes cell count, well-grna, and well-gene plots to disk.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If required keys are missing or regression type is invalid.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.process_reads">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">process_reads</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">csv_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_threshold</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#process_reads"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.process_reads" title="Link to this definition"></a></dt>
<dd><p>Process barcode count data and compute fractional abundance of each gRNA per well.</p>
<p>This function loads a count table, standardizes metadata columns, computes the
relative abundance of each gRNA in a well (<cite>fraction</cite>), and filters based on a
minimum threshold. It also parses <cite>grna</cite> strings into gene-level information if
applicable.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>csv_path</strong> (<em>str</em><em> or </em><em>pd.DataFrame</em>) – Path to CSV file or preloaded DataFrame with count data.</p></li>
<li><p><strong>fraction_threshold</strong> (<em>float</em>) – Minimum fraction of reads per well for inclusion.</p></li>
<li><p><strong>plate</strong> (<em>str</em>) – Plate identifier to use if not present in metadata.</p></li>
<li><p><strong>filter_column</strong> (<em>str</em><em> or </em><em>list</em><em>, </em><em>optional</em>) – Columns to filter values from.</p></li>
<li><p><strong>filter_value</strong> (<em>str</em><em> or </em><em>list</em><em>, </em><em>optional</em>) – Values to exclude from filter_column(s).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Filtered DataFrame with columns [‘prc’, ‘grna’, ‘fraction’].</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If required metadata columns are missing.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.apply_transformation">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">apply_transformation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transform</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#apply_transformation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.apply_transformation" title="Link to this definition"></a></dt>
<dd><p>Apply a mathematical transformation to a variable.</p>
<p>Supported transformations include logarithm, square root, and square.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>X</strong> (<em>array-like</em>) – Input variable or DataFrame column.</p></li>
<li><p><strong>transform</strong> (<em>str</em>) – Type of transformation (‘log’, ‘sqrt’, or ‘square’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Transformer object, or None if transform is unrecognized.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>FunctionTransformer or None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.check_normality">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">check_normality</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">variable_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#check_normality"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.check_normality" title="Link to this definition"></a></dt>
<dd><p>Check if data follows a normal distribution using the Shapiro-Wilk test.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>array-like</em>) – Data to test for normality.</p></li>
<li><p><strong>variable_name</strong> (<em>str</em>) – Name of the variable (used in print statements).</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – Whether to print test results and interpretation.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>True if data is normally distributed (p &gt; 0.05), False otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.clean_controls">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">clean_controls</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">values</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#clean_controls"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.clean_controls" title="Link to this definition"></a></dt>
<dd><p>Remove rows from a DataFrame based on specified control values in a given column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – Input DataFrame.</p></li>
<li><p><strong>values</strong> (<em>list</em>) – List of control values to remove.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – Column from which to remove control values.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Filtered DataFrame with control rows removed.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.process_scores">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">process_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependent_variable</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plate</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_cell_count</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">agg_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'mean'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">transform</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regression_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ols'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#process_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.process_scores" title="Link to this definition"></a></dt>
<dd><p>Aggregate and transform single-cell phenotype scores for regression input.</p>
<p>Groups phenotypic measurements by well (<cite>prc</cite>) and applies optional aggregation
(mean, median, quantile). Filters out wells with low cell counts. Optionally applies
mathematical transformation and checks for normality of the dependent variable.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – Input DataFrame with single-cell phenotype scores.</p></li>
<li><p><strong>dependent_variable</strong> (<em>str</em>) – Name of the phenotype column to model.</p></li>
<li><p><strong>plate</strong> (<em>str</em>) – Plate ID to assign if not present in metadata.</p></li>
<li><p><strong>min_cell_count</strong> (<em>int</em>) – Minimum number of cells per well to retain.</p></li>
<li><p><strong>agg_type</strong> (<em>str</em>) – Aggregation method (‘mean’, ‘median’, ‘quantile’, None).</p></li>
<li><p><strong>transform</strong> (<em>str</em><em> or </em><em>None</em>) – Optional transformation (‘log’, ‘sqrt’, or ‘square’).</p></li>
<li><p><strong>regression_type</strong> (<em>str</em>) – Type of regression (affects aggregation logic for ‘poisson’).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>(aggregated_df, dependent_variable_name)</dt><dd><ul class="simple">
<li><p>aggregated_df: Aggregated and filtered phenotype DataFrame.</p></li>
<li><p>dependent_variable_name: Possibly transformed column name used for modeling.</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If required metadata columns are missing or aggregation type is invalid.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.generate_ml_scores">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">generate_ml_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">settings</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#generate_ml_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.generate_ml_scores" title="Link to this definition"></a></dt>
<dd><p>Perform machine learning analysis on single-cell measurement data from one or more screen sources.</p>
<p>This function loads and merges measurement databases, computes additional features (e.g., recruitment
scores, shortest distance between compartments), handles annotations (for supervised classification),
trains a machine learning model (e.g., XGBoost), computes SHAP values, and generates output files
including predictions, importance scores, SHAP plots, and well-level heatmaps.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>settings</strong> (<em>dict</em>) – Dictionary of analysis parameters and options. Required keys include:
- ‘src’ (str or list): Path(s) to screen folders with measurement databases.
- ‘channel_of_interest’ (int): Channel index used for computing recruitment scores.
- ‘location_column’ (str): Column used to group wells for training (e.g., ‘columnID’).
- ‘positive_control’ (str): Value of <cite>location_column</cite> representing positive control wells.
- ‘negative_control’ (str): Value of <cite>location_column</cite> representing negative control wells.
- ‘exclude’ (list): List of feature names to exclude.
- ‘n_repeats’ (int): Number of repetitions for permutation importance.
- ‘top_features’ (int): Number of top features to keep and plot.
- ‘reg_alpha’, ‘reg_lambda’ (float): Regularization strength for XGBoost.
- ‘learning_rate’ (float): Learning rate for XGBoost.
- ‘n_estimators’ (int): Number of boosting rounds for the model.
- ‘test_size’ (float): Fraction of labeled data to use for testing.
- ‘model_type_ml’ (str): Type of model to use (‘xgboost’, ‘random_forest’, etc.).
- ‘n_jobs’ (int): Number of parallel jobs.
- ‘remove_low_variance_features’ (bool): Whether to drop near-constant features.
- ‘remove_highly_correlated_features’ (bool): Whether to drop redundant features.
- ‘prune_features’ (bool): Whether to perform SelectKBest-based pruning.
- ‘cross_validation’ (bool): Whether to use 5-fold cross-validation.
- ‘heatmap_feature’ (str): Feature to plot as a plate heatmap.
- ‘grouping’ (str): Column used to group wells in the heatmap (e.g., ‘prc’).
- ‘min_max’ (tuple): Min/max range for heatmap normalization.
- ‘minimum_cell_count’ (int): Minimum number of cells per well for heatmap inclusion.
- ‘cmap’ (str): Colormap for the heatmap.
- ‘save_to_db’ (bool): If True, update the database with prediction column.
- ‘verbose’ (bool): If True, print detailed logs.
- ‘annotation_column’ (str or None): If provided, annotate data from the <cite>png_list</cite> table.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>output (list): Includes DataFrames and model objects from the <cite>ml_analysis</cite> function.</p></li>
<li><p>plate_heatmap (matplotlib.Figure): Plate-level heatmap of the selected feature.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If required columns are missing or specified features are not found.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.ml_analysis">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">ml_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channel_of_interest</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">location_column</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'columnID'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">positive_control</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'c2'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">negative_control</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'c1'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_repeats</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">top_features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">30</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reg_alpha</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reg_lambda</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">learning_rate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1e-05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_estimators</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">test_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'xgboost'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_jobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_low_variance_features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_highly_correlated_features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prune_features</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cross_validation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#ml_analysis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.ml_analysis" title="Link to this definition"></a></dt>
<dd><p>Train a machine learning classifier to distinguish between positive and negative control wells,
compute feature importance via permutation and model-based methods, and assign predictions to all rows.</p>
<p>This function supports several classifier types (XGBoost, random forest, logistic regression, etc.),
with options for feature selection, cross-validation, and SHAP-ready output. It returns predictions,
probability estimates, and feature importances, and computes classification metrics.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pd.DataFrame</em>) – Input DataFrame with features and metadata.</p></li>
<li><p><strong>channel_of_interest</strong> (<em>int</em>) – Channel index used for filtering relevant features.</p></li>
<li><p><strong>location_column</strong> (<em>str</em>) – Column that defines grouping (e.g., ‘columnID’) to select control wells.</p></li>
<li><p><strong>positive_control</strong> (<em>str</em><em> or </em><em>list</em>) – Identifier(s) in <cite>location_column</cite> for positive control group.</p></li>
<li><p><strong>negative_control</strong> (<em>str</em><em> or </em><em>list</em>) – Identifier(s) in <cite>location_column</cite> for negative control group.</p></li>
<li><p><strong>exclude</strong> (<em>list</em><em> or </em><em>str</em><em>, </em><em>optional</em>) – Feature names or substrings to exclude from analysis.</p></li>
<li><p><strong>n_repeats</strong> (<em>int</em>) – Number of repetitions for permutation importance calculation.</p></li>
<li><p><strong>top_features</strong> (<em>int</em>) – Number of top features to retain and visualize.</p></li>
<li><p><strong>reg_alpha</strong> (<em>float</em>) – L1 regularization term (only for XGBoost).</p></li>
<li><p><strong>reg_lambda</strong> (<em>float</em>) – L2 regularization term (only for XGBoost).</p></li>
<li><p><strong>learning_rate</strong> (<em>float</em>) – Learning rate for gradient-based models (e.g., XGBoost).</p></li>
<li><p><strong>n_estimators</strong> (<em>int</em>) – Number of trees or boosting iterations.</p></li>
<li><p><strong>test_size</strong> (<em>float</em>) – Fraction of labeled data used for testing.</p></li>
<li><p><strong>model_type</strong> (<em>str</em>) – Model type to train: ‘xgboost’, ‘random_forest’, ‘logistic_regression’, or ‘gradient_boosting’.</p></li>
<li><p><strong>n_jobs</strong> (<em>int</em>) – Number of CPU cores to use (where supported).</p></li>
<li><p><strong>remove_low_variance_features</strong> (<em>bool</em>) – If True, drop near-constant features.</p></li>
<li><p><strong>remove_highly_correlated_features</strong> (<em>bool</em>) – If True, drop features with high pairwise correlation.</p></li>
<li><p><strong>prune_features</strong> (<em>bool</em>) – If True, apply SelectKBest to reduce features before training.</p></li>
<li><p><strong>cross_validation</strong> (<em>bool</em>) – If True, use 5-fold stratified cross-validation instead of single split.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – If True, print detailed logs and show diagnostic plots.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><dl class="simple">
<dt>[0] df (pd.DataFrame): Original DataFrame with added predictions, probabilities, and metadata.</dt><dd><ul>
<li><p>[1] permutation_df (pd.DataFrame): Permutation importance for top features.</p></li>
<li><p>[2] feature_importance_df (pd.DataFrame): Model-based feature importance scores (if supported).</p></li>
<li><p>[3] model (sklearn-compatible): Trained model object.</p></li>
<li><p>[4] X_train (pd.DataFrame): Training feature matrix.</p></li>
<li><p>[5] X_test (pd.DataFrame): Test feature matrix.</p></li>
<li><p>[6] y_train (pd.Series): Training target labels.</p></li>
<li><p>[7] y_test (pd.Series): Test target labels.</p></li>
<li><p>[8] metrics_df (pd.DataFrame): Summary classification metrics (precision, recall, f1).</p></li>
<li><p>[9] features (list): Final list of features used in training.</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
<dl class="simple">
<dt>list:</dt><dd><ul class="simple">
<li><p>[0] permutation_fig (matplotlib.Figure): Bar plot of permutation importances.</p></li>
<li><p>[1] feature_importance_fig (matplotlib.Figure): Bar plot of model importances.</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.shap_analysis">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">shap_analysis</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">model</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X_train</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">X_test</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#shap_analysis"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.shap_analysis" title="Link to this definition"></a></dt>
<dd><p>Compute and return a SHAP summary plot figure for a trained model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>model</strong> – A trained machine learning model compatible with SHAP.</p></li>
<li><p><strong>X_train</strong> (<em>pd.DataFrame</em>) – Training features used to initialize the SHAP explainer.</p></li>
<li><p><strong>X_test</strong> (<em>pd.DataFrame</em>) – Test features for which SHAP values will be computed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>SHAP summary plot figure.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>matplotlib.figure.Figure</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.find_optimal_threshold">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">find_optimal_threshold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">y_true</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_pred_proba</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#find_optimal_threshold"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.find_optimal_threshold" title="Link to this definition"></a></dt>
<dd><p>Determine the optimal decision threshold that maximizes the F1-score
based on predicted probabilities.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>y_true</strong> (<em>array-like</em>) – Ground-truth binary labels.</p></li>
<li><p><strong>y_pred_proba</strong> (<em>array-like</em>) – Predicted class probabilities (positive class).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Threshold that yields the highest F1-score.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="spacr.ml.interperate_vision_model">
<span class="sig-prename descclassname"><span class="pre">spacr.ml.</span></span><span class="sig-name descname"><span class="pre">interperate_vision_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">settings</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../../../_modules/spacr/ml.html#interperate_vision_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#spacr.ml.interperate_vision_model" title="Link to this definition"></a></dt>
<dd><p>Perform feature interpretation on vision model predictions using feature importance,
permutation importance, and SHAP analysis.</p>
<dl class="simple">
<dt>Steps:</dt><dd><ol class="arabic simple">
<li><p>Loads and merges measurement and score data.</p></li>
<li><p>Computes Random Forest feature importances.</p></li>
<li><p>Computes permutation-based feature importances.</p></li>
<li><p>Runs SHAP analysis on selected features.</p></li>
<li><p>Aggregates SHAP values by compartment and channel and visualizes them using radar plots.</p></li>
</ol>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>settings</strong> (<em>dict</em>) – Dictionary containing configuration options, including:
- src (str): Path to measurement database.
- tables (list): List of tables to include.
- score_column (str): Column with predicted or measured scores.
- top_features (int): Number of features to retain and plot.
- shap (bool): Whether to perform SHAP analysis.
- shap_sample (bool): Whether to subsample data before SHAP.
- feature_importance (bool): Whether to compute Random Forest importance.
- permutation_importance (bool): Whether to compute permutation importance.
- n_jobs (int): Number of parallel jobs for model training and permutation.
- save (bool): Whether to save result CSVs.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Merged and scored dataset used for interpretation.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../measure/index.html" class="btn btn-neutral float-left" title="spacr.measure" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../plot/index.html" class="btn btn-neutral float-right" title="spacr.plot" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright .</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>